lib/route-recognizer/normalizer.js

```diff
@@ -74,9 +74,52 @@ function normalizeSegment(segment) {
   return decodeURIComponentExcept(segment, reservedHex);
 }
 
+function encodeURIComponentExcept(string, reservedChars) {
+  var pieces       = [];
+  var separators   = [];
+  var currentPiece = '';
+
+  for (var idx=0; idx < string.length; idx++) {
+    var char = string[idx];
+    if (reservedChars.indexOf(char) === -1) {
+      currentPiece += char;
+    } else {
+      pieces.push(currentPiece);
+      separators.push(char);
+      currentPiece = '';
+    }
+  }
+  if (currentPiece.length) {
+    pieces.push(currentPiece);
+  }
+
+  pieces = pieces.map(encodeURIComponent);
+  var ret = '';
+  while (pieces.length) {
+    ret += pieces.shift();
+    if (separators.length) {
+      ret += separators.shift();
+    }
+  }
```

Comment:
You can do this without modification of the arrays and without an `if` which I believe will be faster:
```js
separators.push('');
for (var i = 0; i < pieces.length; i++) {
  ret = ret + pieces[i] + separators[i];
}
```